### PR TITLE
RATIS-2028. Refactor RaftLog to supply log as ReferenceCountedObject

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

Refactor RaftLog to supply log as ReferenceCountedObject

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-2028

## How was this patch tested?

CI.

### CODE REVIEW

**Comment by duongkame** [comment]
cc. @szetszwo

File: ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java [hunk 1]
```diff
@@ -58,8 +59,21 @@ default boolean contains(TermIndex ti) {
...
+    return ref != null ? ref.get() : null;
+  }
+
+  /**
+   * @return a ReferenceCountedObject to the  log entry corresponding to the given log index if it exists.
+   * otherwise, null.
+   */
+  default ReferenceCountedObject<LogEntryProto> getWithRef(long index) throws RaftLogIOException {
+    return ReferenceCountedObject.wrap(get(index));
+  }
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java:76]: Similar to `TransactionContext.getLogEntryRef()` below.  We should think about if the caller should invoke `retain()` before using it.  If `retain()` is not needed, when is it safe to call `get()`.

https://github.com/apache/ratis/blob/1bd451046b2d33fe063712d7e811262bef580a4e/ratis-server-api/src/main/java/org/apache/ratis/statemachine/TransactionContext.java#L115-L130

**Review by szetszwo — COMMENTED** [review]
@duongkame , thanks for working not this!  We should talk about the contract of the new `RaftLog.getWithRef`.   Are `retain()`/`release()` required?  If not, when is safe to use it?

Currently, `SegmentedRaftLog` may release the returned `ref` and causing an exception as below:
1.  SegmentedRaftLog loads an entry, creates `ref` and calls `retain()` in order to save it to cache.
2. SegmentedRaftLog.getWithRef(..) returns `ref` to a caller.
3. SegmentedRaftLog cache evicts the entry, the reference count becomes 0 and `ref` is completely released.
4. The caller use `ref` by calling `get()` or `retain()`.  Then, it will get an exception.

File: ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java (hunk 1 above)
Comment (duongkame) [reply ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java:76]: `TransactionContext.getLogEntryRef()` will be protected by a retain/release pair from the framework. So, the StateMachine code doesn't need to call retain/release unless extending the LogEntry usage outside the method scope. It's where the javadoc above from.

```
// in StateMachineUpdater.
final ReferenceCountedObject<LogEntryProto> next = raftLog.getWithRef(nextIndex);
next.retain();
try {
  server.applyLogToStateMachine(next);
} finally {
  next.release();
}
```

**Comment by duongkame** [comment]
> @duongkame , thanks for working not this! We should talk about the contract of the new `RaftLog.getWithRef`. Are `retain()`/`release()` required? If not, when is safe to use it?
> 
> Currently, `SegmentedRaftLog` may release the returned `ref` and causing an exception as below:
> 
> 1. SegmentedRaftLog loads an entry, creates `ref` and calls `retain()` in order to save it to cache.
> 2. SegmentedRaftLog.getWithRef(..) returns `ref` to a caller.
> 3. SegmentedRaftLog cache evicts the entry, the reference count becomes 0 and `ref` is completely released.
> 4. The caller use `ref` by calling `get()` or `retain()`.  Then, it will get an exception.

Yeah, that's a good point. I had same question but wasn't sure if the log entry can be evicted shortly after being called with `get` from SegmentedRaftLog.  If theoretically eviction can happen immediately, calling `retain()` after `getWithRef()` can help either as there's no coordination between the SegmentedRaftLog internal state and the `ReferenceCountedObject` after it is returned.

To properly prevent this error, the `retain()` needs to be called in the readLock before `getWithRef` returns to prevent it from being cleaned up. It means a new API that looks like

```
# SegmentedRaftLog API
void readLogEntry(long indec, Consumer<LogEntryProto> consumer) {
   try (AutoCloseableLock readLock = readLock()) {
     segment = cache.getSegment(index);
     final ReferenceCountedObject<LogEntryProto> entry = segment.getEntryFromCache(record.getTermIndex());
     LogEntryProto log = entry.retain();
     try {
       consumer.consume(log);
     } finally {
       entry.release();
     }
   }
}


```

**Comment by szetszwo** [comment]
Could we retain it before returns and ask the caller to release it?
```java
//RaftLog
   /**
-   * @return a ReferenceCountedObject to the  log entry corresponding to the given log index if it exists.
-   * otherwise, null.
+   * @return a retained {@link ReferenceCountedObject} to the log entry corresponding to the given index
+   *         if it exists; otherwise, return null.
+   *         Since the returned reference is retained,
+   *         the caller must call {@link ReferenceCountedObject#release()}} after use.
    */
-  default ReferenceCountedObject<LogEntryProto> getWithRef(long index) throws RaftLogIOException {
+  default ReferenceCountedObject<LogEntryProto> retainEntry(long index) throws RaftLogIOException {
     return ReferenceCountedObject.wrap(get(index));
   }
```

```java
@@ -287,6 +287,7 @@ public final class SegmentedRaftLog extends RaftLogBase {
       }
       final ReferenceCountedObject<LogEntryProto> entry = segment.getEntryFromCache(record.getTermIndex());
       if (entry != null) {
+        entry.retain();
         getRaftLogMetrics().onRaftLogCacheHit();
         return entry;
       }
```

**Comment by duongkame** [comment]
> Could we retain it before returns and ask the caller to release it?
> 

Sounds good. We can call `retain` in the readLock.

**Review by szetszwo — COMMENTED** [review]
@duongkame , thanks for the update!  Please see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13067355/1045_review.patch

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java [hunk 2]
```diff
@@ -284,25 +303,39 @@ record = segment.getLogRecord(index);
...
 
     // the entry is not in the segment's cache. Load the cache without holding the lock.
     getRaftLogMetrics().onRaftLogCacheMiss();
     cacheEviction.signal();
-    return segment.loadCache(record);
+    ReferenceCountedObject<LogEntryProto> loaded = segment.loadCache(record);
+    if (loaded != null) {
+      loaded.retain();
+    }
+    return loaded;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java]: We should revert this part, i.e.
```java
@@ -301,11 +314,7 @@ public final class SegmentedRaftLog extends RaftLogBase {
     // the entry is not in the segment's cache. Load the cache without holding the lock.
     getRaftLogMetrics().onRaftLogCacheMiss();
     cacheEviction.signal();
-    ReferenceCountedObject<LogEntryProto> loaded = segment.loadCache(record);
-    if (loaded != null) {
-      loaded.retain();
-    }
-    return loaded;
+    return segment.loadCache(record);
   }
```

**Review by szetszwo — COMMENTED** [review]
@duongkame , we forgot to revert the `retain()` call of the loaded segment; see the comment inlined.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java (hunk 2 above)
Comment (duongkame) [reply ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java]: think I did. https://github.com/apache/ratis/pull/1045/commits/1414bd63fd10ea27aa9699c71bb9d7ebc8dbb803.

Maybe it was pushed after you'd reviewed, sorry.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**